agents/temporal_difference.c

```diff
@@ -0,0 +1,120 @@
+#include <float.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "game.h"
+#include "temporal_difference.h"
+#include "util.h"
+
+void init_td_agent(td_agent_t *agent, unsigned int state_num, char player)
+{
+    (*agent).player = player;
+    (*agent).state_value = malloc(sizeof(float) * state_num);
+    if (!((*agent).state_value))
+        exit(1);
+    for (unsigned int i = 0; i < state_num; i++)
+        (*agent).state_value[i] = get_score(hash_to_table(i), player);
+}
+
+void load_model(td_agent_t *agent,
+                unsigned int state_num,
+                const char *model_path)
+{
+    FILE *fptr = NULL;
+    if ((fptr = fopen(model_path, "rb")) == NULL) {
+        printf("could not open state value table, train first\n");
+        exit(1);
+    }
+    long offset = ((*agent).player == 'O') ? 0 : state_num * sizeof(float);
+    fseek(fptr, offset, SEEK_SET);
+    if (fread((*agent).state_value, state_num * sizeof(float), 1, fptr) != 1) {
+        printf("could not load\n");
+        exit(1);
+    }
+    fclose(fptr);
+}
+
+int *get_available_moves(char *table, int *ret_size)
+{
+    int move_cnt = 0;
+    int *available_moves = malloc(N_GRIDS * sizeof(int));
+    if (!available_moves)
+        exit(1);
+    for (int i = 0; i < N_GRIDS; i++) {
+        if (table[i] == ' ') {
+            available_moves[move_cnt++] = i;
+        }
+    }
+    *ret_size = move_cnt;
+    return available_moves;
+}
+
+int get_action_exploit(char *table, td_agent_t *agent)
+{
+    int move_cnt = 0;
+    int *available_moves = get_available_moves(table, &move_cnt);
+    int max_act = -1;
+    float max_q = -DBL_MAX;
```

Comment:
Since single-precision floating-point number is used here, we should use ```FLT_MAX``` instead of ```DBL_MAX```.